[#947] Use `utfEncode'` to log bytestrings

korayal opened the pull request:
closes #947

endgame commented on lib/amazonka-core/src/Amazonka/Data/Log.hs:
```diff
@@ -34,10 +34,14 @@ instance ToLog ByteStringBuilder where
 instance ToLog ByteStringLazy where
-  build = Build.lazyByteString
+  build lbs = case LText.decodeUtf8' lbs of
+    Left _ -> "non-printable " <> build (LBS.length lbs) <> " lazy bytes"
```
```suggestion
    Left _ -> mconcat ["non-printable lazy ByteString (" <> build (LBS.length lbs) <> " bytes)"]
```
And similarly below.

endgame reviewed (commented):
Can you please add an entry to `lib/amazonka/CHANGELOG.md`, which is the master changelog for the project?

Also, are we sure that this is the correct check to make? What if the user's terminal is set to some other encoding scheme? A TextEncoding (as returned by `System.IO.hGetEncoding`) isn't even in `Eq`, so I'm not sure what the best way is to fix this. I suspect what we have here might be a bit too naive; do you have any thoughts about other ways we could check this? It may also be the case that we successfully decode UTF-8 but it's got non-printable characters in it; do we want to do that check here too?

In the medium term, I probably want to extract logging from the main amazonka library so that people can use whatever logging framework suits them. Then handling non-printable stuff becomes the log library's concern.

korayal commented:
> Can you please add an entry to lib/amazonka/CHANGELOG.md, which is the master changelog for the project?

I've updated the changelog and added a section `Unreleased` at the top. would that work?

korayal commented:
> Also, are we sure that this is the correct check to make? What if the user's terminal is set to some other encoding scheme? A TextEncoding (as returned by System.IO.hGetEncoding) isn't even in Eq, so I'm not sure what the best way is to fix this. I suspect what we have here might be a bit too naive; do you have any thoughts about other ways we could check this? It may also be the case that we successfully decode UTF-8 but it's got non-printable characters in it; do we want to do that check here too?

I've added another check with `isPrint`, where it should ensure that whatever we're trying to print contains printable characters. At this point I think assuming the request body (if not compressed) is a valid utf-8 text since it's what we're sending to AWS. The part that prints it out is still left to `Build.lazyByteString` / `Build.byteString`, so at worst case we're keeping the existing behavior.

> In the medium term, I probably want to extract logging from the main amazonka library so that people can use whatever logging framework suits them. Then handling non-printable stuff becomes the log library's concern.

that'd be great.

endgame commented on lib/amazonka-core/src/Amazonka/Data/Log.hs:
```diff
@@ -34,10 +35,14 @@ instance ToLog ByteStringBuilder where
 instance ToLog ByteStringLazy where
-  build = Build.lazyByteString
+  build lbs = case LText.decodeUtf8' lbs of
+    Right t | all Char.isPrint (LText.unpack t) -> Build.lazyByteString lbs
```
`text` defines its own versions of `all`, so you don't need to unpack.

korayal replied:
replaced. thanks!

endgame commented on lib/amazonka/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## Unreleased
+
+- `amazonka`: During DEBUG logging, do not print bytestrings that have unprintable characters [\#952](https://github.com/brendanhay/amazonka/pull/952)
```
Please put the PR link on the next line to match existing entries. I'd also remove the word "DEBUG", since we're fixing this by changing typeclasses.

korayal replied:
done.

korayal commented:
@endgame are we OK with the latest state of this PR?

endgame commented on lib/amazonka-core/src/Amazonka/Data/Log.hs:
```diff
@@ -34,10 +35,14 @@ instance ToLog ByteStringBuilder where
-  build = Build.byteString
+  build bs = case Text.decodeUtf8' bs of
+    Right t | Text.all Char.isPrint t -> Build.byteString bs
+    _ -> mconcat ["non-printable strict ByteString (", build (BS.length bs), ")"]
```
```suggestion
    _ -> mconcat ["non-printable strict ByteString (", build (BS.length bs), " bytes)"]
```

endgame commented on lib/amazonka-core/src/Amazonka/Data/Log.hs:
```diff
@@ -34,10 +35,14 @@ instance ToLog ByteStringBuilder where
-  build = Build.lazyByteString
+  build lbs = case LText.decodeUtf8' lbs of
+    Right lt | LText.all Char.isPrint lt -> Build.lazyByteString lbs
+    _ -> mconcat ["non-printable lazy ByteString (", build (LBS.length lbs), ")"]
```
```suggestion
    _ -> mconcat ["non-printable lazy ByteString (", build (LBS.length lbs), " bytes)"]
```

endgame reviewed (commented):
Thank you for this, and sorry for the delay.
